### PR TITLE
use async-retry to not handle/redo the poll-timing logic

### PR DESCRIPTION
**Pull request by louis-bompart**
imho, this is a solved problem in the field, and we'd reinvent the wheel by not using a package to do it (it would be kinda an anti-pattern of how NPM works too)

### CODE REVIEW

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 1]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
     });
   }
 
-  public async waitUntilDone(
-    operationToWaitFor?: ResourceSnapshotsReportType | null,
-    options: WaitUntilDoneOptions = {},
-    iteratee = (_report: ResourceSnapshotsReportModel) => {}
+  public waitUntilDone(
+    operationToWaitFor: ResourceSnapshotsReportType | null,
+    options: Partial<WaitUntilDoneOptions> = {},
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:144]: Given the spread l145, we can provide a `Partial`. This allow the caller to only set the options he want

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 2]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
...
-        timeout = setTimeout(() => {
-          const err = new SnapshotOperationTimeoutError(this);
-          clearInterval(interval);
-          reject(err);
-        }, toMilliseconds(opts.wait));
+
+    return retry(
+      this.waitUntilDoneRetryFunction(operationToWaitFor, onRetryCb),
+      {
+        retries: Infinity,
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:153]: We don't care about the number of attempt, so `Infinity`

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 3]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
...
+        minTimeout: toMilliseconds(opts.waitInterval),
+        maxTimeout: toMilliseconds(opts.waitInterval),
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:155]: We want to retry every `opts.waitInterval` seconds.
Setting both limits of the Timeout ensure that this
`Math.min(random * minTimeout * Math.pow(factor, attempt), maxTimeout)` (see [node-retry](https://github.com/tim-kos/node-retry)) will always be eq to `opts.waitInterval`
given that: 
```
random=1 // i.e. no random
factor=2 // default value
attempt>=0 // 
```
 => Math.pow(factor,attempt)>=1
 => lower bound of the Math.min>=opts.waitInterval.
 => Math.min === opts.waitInterval

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 4]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
...
-          clearInterval(interval);
-          resolve();
-        }
-      }, toMilliseconds(opts.waitInterval));
-    });
+      if (isOngoing || !isExpectedOperation) {
+        throw new SnapshotOperationTimeoutError(this);
+      }
+      return;
+    }).bind(this);
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:181]: extracted the function in a factory to keep things flat and simple tor ead

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 5]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
...
+      onRetryCb(this.latestReport);
 
-        if (!isOngoing && isExpectedOperation) {
-          clearTimeout(timeout);
-          clearInterval(interval);
-          resolve();
-        }
-      }, toMilliseconds(opts.waitInterval));
-    });
+      if (isOngoing || !isExpectedOperation) {
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:177]: Reversed the condition using De Morgan Law to keep it simple. No return needed in the end

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 6]
```diff
@@ -138,45 +139,45 @@ export class Snapshot {
...
-
-    await new Promise<void>((resolve, reject) => {
-      if (opts.wait > 0) {
-        timeout = setTimeout(() => {
-          const err = new SnapshotOperationTimeoutError(this);
-          clearInterval(interval);
-          reject(err);
-        }, toMilliseconds(opts.wait));
+
+    return retry(
```
Comment (louis-bompart) [inline packages/cli/src/lib/snapshot/snapshot.ts:150]: We can return the retry straight away and thus remove the async from waitUntilDone and just use the promise created by `async-retry`

File: packages/cli/src/lib/snapshot/snapshot.ts (hunk 2 above)
Comment (y-lakhdar) [reply packages/cli/src/lib/snapshot/snapshot.ts:153]: Are you sure it supports Infinity? And that you don't have to use `Number.MAX_SAFE_INTEGER`

**Review by y-lakhdar — APPROVED** [review]
Clean

**Review by y-lakhdar — APPROVED** [review]

File: packages/cli/src/lib/snapshot/snapshot.ts [hunk 7]
```diff
@@ -138,45 +139,44 @@ export class Snapshot {
...
-          clearTimeout(timeout);
-          clearInterval(interval);
-          resolve();
-        }
-      }, toMilliseconds(opts.waitInterval));
-    });
+      if (isOngoing || !isExpectedOperation) {
+        throw new SnapshotOperationTimeoutError(this);
+      }
+    }).bind(this);
```
Comment (y-lakhdar) [inline packages/cli/src/lib/snapshot/snapshot.ts:180]: why bind(this)

**Review by y-lakhdar — APPROVED** [review]

**Pull request merged**